src/libraries/System.Runtime.Serialization.BinaryFormat/src/Records/ArraySingleObjectRecord.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection.Metadata;
+
+namespace System.Runtime.Serialization.BinaryFormat;
+
+/// <summary>
+///  Single dimensional array of objects.
+/// </summary>
+/// <remarks>
+///  <para>
+///   <see href="https://learn.microsoft.com/openspecs/windows_protocols/ms-nrbf/982b2f50-6367-402a-aaf2-44ee96e2a5e0">
+///    [MS-NRBF] 2.4.3.2
+///   </see>
+///  </para>
+/// </remarks>
+internal sealed class ArraySingleObjectRecord : ArrayRecord<object?>
+{
+    private static TypeName? s_elementTypeName;
+
+    private ArraySingleObjectRecord(ArrayInfo arrayInfo) : base(arrayInfo) => Records = [];
+
+    public override RecordType RecordType => RecordType.ArraySingleObject;
+
+    public override TypeName ElementTypeName
+        => s_elementTypeName ??= TypeName.Parse(typeof(object).FullName.AsSpan()).WithCoreLibAssemblyName();
+
+    private List<SerializationRecord> Records { get; }
+
+    public override bool IsTypeNameMatching(Type type) => type == typeof(object[]);
```

Comment:
This does not implement what the method name says. The method name says that it should match type names, but the implementation matches the exact runtime type identities.

For example, if this gets method passed Type for object array that came from `System.Reflection.MetadataLoadContext`, it will return false - even though the type name is going to match.

Reply:
`ArraySingleObjectRecord` is produced only when a single-dimension, zero-index array of `System.Object` gets serialized (`object[]`). This method does not check the type name (same goes for arrays of strings and few other records) because it's more expensive than just checking the type against hardcoded input.

Reply:
I understand that the `typeof(object)` shortcut is faster when operating with runtime provided System.Type implementation. My point is that it is not correct - the method does not do what its name suggests in all cases.

I think the payload reader should be designed to work against any System.Type implementation, and not just the runtime provided one, so that it can be used without loading the types involved for execution.

Reply:
@jkotas Thanks for the explanation, I had no idea people can create custom types that derive from `Type`. But even if they do, I doubt that they could produce `ArraySingleObjectRecord` for such a type, as `BinaryFormatter` is most likely having a very similar check in its serializer code.

Reply:
> I had no idea people can create custom types that derive from Type

Yes, it has been possible for a very long time (since ~2010). 

It allows people to use the familiar System.Reflection APIs to implement libraries and tools like compilers against one .NET TFM and allow them to target (read or produce binaries for) a different TFM that can be passed in as an input to them. These tools never execute code from the target TFM, they just inspect the APIs available in the TFM, etc.

The low-level NRBF reader seems to be a library like that. It goes to great lengths to avoid loading target types for execution. There are just a few outliers like this one place where it hardcodes the runtime types instead of going through the abstraction. It may be worth it fix it to go through the proper abstraction to allow it to be used in more configurations.